### PR TITLE
Claymore Cameras + Alerts and Better Camera Names - WIP

### PR DESCRIPTION
**Pull request by Surrealaser**
:cl: Surrealistik
add: Claymores now have cameras and provide alerts when tripped.
tweak: Sentry/claymore cameras now feature information about their area and coordinates. This is updated whenever they're powered on.
fix: turned off alerts/safety for pre-made 'dumb' sentries
/:cl:

### CODE REVIEW

**Comment by Surrealaser** [comment]
> Where are these cameras supposed to be watched through?

The military network that command has access to.

File: code/game/machinery/camera/camera.dm [hunk 1]
```diff
@@ -8,7 +8,7 @@
 	active_power_usage = 10
 	layer = FLY_LAYER
 
-	var/list/network = list("military")
+	var/list/network = list("LEADER")
```
Comment (Rohesie) [inline code/game/machinery/camera/camera.dm]: Why the name change? And why the caps lock?

Comment (Surrealaser) [reply code/game/machinery/camera/camera.dm]: Because that's the network name; I didn't select it.

**Comment by Surrealaser** [comment]
@Rohesie I'm aware, unfortunately my time today is limited.

**Comment by DominikPanic** [comment]
See Rohesie's requested changes.

**Comment by DominikPanic** [comment]
Also conflicts

File: code/game/objects/items/explosives/mine.dm [hunk 2]
```diff
@@ -170,4 +178,20 @@
...
+	if(!M)
+		return
+	var/notice = "<b>ALERT! [src] detonated. Hostile/unknown: [M] Detected at: [get_area(M)]. Coordinates: (X: [M.x], Y: [M.y]).</b>"
+	playsound(loc, 'sound/machines/warning-buzzer.ogg', 50, FALSE)
+	var/mob/living/silicon/ai/AI = new/mob/living/silicon/ai(src, null, null, 1)
+	AI.SetName("Smartmine Alert System")
+	AI.aiRadio.talk_into(AI,"[notice]","Theseus","announces")
+	qdel(AI)
+
+	switch(trigger_type) //Makes sure we announce first before detonation.
```
Comment (Rohesie) [inline code/game/objects/items/explosives/mine.dm]: Does this require a `switch`, if there's a single option?
Could be simplified with `if(trigger_type == "explosive" && tripwire)`

**Comment by Rohesie** [comment]
This has been tested, but has change requests pending.

File: code/modules/projectiles/updated_projectiles/ammo_datums.dm [hunk 3]
```diff
@@ -640,6 +640,10 @@
 	penetration	= CONFIG_GET(number/combat_define/high_armor_penetration)
 	bonus_projectiles_amount = CONFIG_GET(number/combat_define/low_proj_extra)
 
+/datum/ammo/bullet/shotgun/flechette/claymore/New()
+	..()
```
Comment (DominikPanic) [inline code/modules/projectiles/updated_projectiles/ammo_datums.dm]:
```suggestion
	. = ..()
```

File: code/modules/cm_marines/equipment/sentries.dm [hunk 4]
```diff
@@ -284,9 +285,10 @@
 	spark_system.set_up(5, 0, src)
 	spark_system.attach(src)
 	cell = new /obj/item/cell/high(src)
+	camera_number = rand(1,1000)
 	camera = new (src)
 	camera.network = list("military")
-	camera.c_tag = "[name] ([rand(0, 1000)])"
+	camera.c_tag = "Sentry: [get_area(src)] | X: [x] | Y: [y]| [camera_number]" //Update Camera name
```
Comment (DominikPanic) [inline code/modules/cm_marines/equipment/sentries.dm]:
```suggestion
	camera.c_tag = "Sentry: [get_area(src)] | X: [x] | Y: [y] | [camera_number]" //Update Camera name
```

File: code/modules/cm_marines/equipment/sentries.dm [hunk 5]
```diff
@@ -284,9 +285,10 @@
 	spark_system.set_up(5, 0, src)
 	spark_system.attach(src)
 	cell = new /obj/item/cell/high(src)
+	camera_number = rand(1,1000)
```
Comment (DominikPanic) [inline code/modules/cm_marines/equipment/sentries.dm]: Eh there could be conflicts but not sure if doing something more complex would be worth it.

File: code/game/objects/items/explosives/mine.dm [hunk 6]
```diff
@@ -170,4 +241,87 @@
...
+	if(isnull(angle))
+		return
+
+	var/muzzle_flash_lum = 3
+
+	SetLuminosity(muzzle_flash_lum)
+	spawn(10)
+		SetLuminosity(-muzzle_flash_lum)
+
+	if(prob(65))
```
Comment (DominikPanic) [inline code/game/objects/items/explosives/mine.dm]: Early return

File: code/game/objects/items/explosives/mine.dm [hunk 7]
```diff
@@ -170,4 +241,87 @@
...
+			qdel(src)
+
+/obj/item/explosive/mine/proc/muzzle_flash(var/angle)
+	if(isnull(angle))
+		return
+
+	var/muzzle_flash_lum = 3
+
+	SetLuminosity(muzzle_flash_lum)
+	spawn(10)
```
Comment (DominikPanic) [inline code/game/objects/items/explosives/mine.dm]: Callback

File: code/game/objects/items/explosives/mine.dm [hunk 8]
```diff
@@ -170,4 +241,87 @@
 		return
 
 	if(linked_claymore && ismob(A))
-		linked_claymore.Bumped(A)
\ No newline at end of file
+		linked_claymore.Bumped(A)
+
+/obj/item/explosive/mine/proc/mine_alert(mob/M)
+	set waitfor = 0
```
Comment (DominikPanic) [inline code/game/objects/items/explosives/mine.dm]:
```suggestion
	set waitfor = FALSE
```

File: code/game/objects/items/explosives/mine.dm [hunk 9]
```diff
@@ -34,6 +37,9 @@
 	if(tripwire)
 		qdel(tripwire)
 		tripwire = null
+	if(camera)
+		qdel(camera)
+		camera = null
```
Comment (DominikPanic) [inline code/game/objects/items/explosives/mine.dm]: `QDEL_NULL`

File: code/game/objects/items/explosives/mine.dm [hunk 10]
```diff
@@ -170,4 +241,87 @@
...
+			playsound(loc, 'sound/weapons/gun_shotgun.ogg', 75, 1)
+			for(var/i = 1 to 5)
+				target = get_turf(target)
+				if (prob(50))
+					var/scatter_x = rand(-1, 1)
+					var/scatter_y = rand(-1, 1)
+					var/turf/new_target = locate(targloc.x + round(scatter_x),targloc.y + round(scatter_y),targloc.z) //Locate an adjacent turf.
+					if(new_target) //Looks like we found a turf.
+						target = new_target
+				sleep(1)
```
Comment (DominikPanic) [inline code/game/objects/items/explosives/mine.dm]: Sleep here is a little iffy. You could callback it.

File: code/game/objects/items/explosives/mine.dm [hunk 11]
```diff
@@ -94,35 +103,95 @@
...
+<B>Current Alarm Mode:</B> [alarm_mode ? "Siren" : "Silent"]<BR>
+<A href='?src=\ref[src];alarm_mode=1'><B>Set Alarm Mode:</B> [alarm_mode ? "Silent" : "Siren"]</A><BR>
+
+</TT>"}
+	user << browse(dat, "window=radio")
+	onclose(user, "radio")
+	return
+
+/obj/item/explosive/mine/Topic(href, href_list)
+	//..()
```
Comment (DominikPanic) [inline code/game/objects/items/explosives/mine.dm]: Commented out code

File: code/game/objects/items/explosives/mine.dm [hunk 12]
```diff
@@ -94,35 +103,95 @@
...
+		else //Multitool allows us to program the claymore otherwise.
+			interface(user)
+
+/obj/item/explosive/mine/proc/interface(mob/living/carbon/user)
+	if(!istype(user))
+		return
+	if(user.mind && user.mind.cm_skills && user.mind.cm_skills.engineer < SKILL_ENGINEER_METAL)
+		user.visible_message("<span class='notice'>[user] fumbles around figuring out how to use [src].</span>",
+		"<span class='notice'>You fumble around figuring out how to use [src].</span>")
+		var/fumbling_time = 20
```
Comment (DominikPanic) [inline code/game/objects/items/explosives/mine.dm]: There is already a lot of defines for fumbling time. Try using one of those.

File: code/modules/cm_marines/equipment/sentries.dm (hunk 5 above)
Comment (Rohesie) [reply code/modules/cm_marines/equipment/sentries.dm]: I made `UNIQUEID` for the laser targets. Could be used for cameras as well.

File: code/game/objects/items/explosives/mine.dm (hunk 9 above)
Comment (DominikPanic) [reply code/game/objects/items/explosives/mine.dm]: QDEL_NULL combines both steps into one, you're now trying to null something that doesn't exist.

File: code/game/objects/items/explosives/mine.dm [hunk 13]
```diff
@@ -94,35 +105,95 @@
 				qdel(tripwire)
 				tripwire = null
 
+		else //Multitool allows us to program the claymore otherwise.
+			interface(user)
+
+/obj/item/explosive/mine/proc/interface(mob/living/carbon/user)
+	if(!istype(user))
+		return
+	if(user.mind && user.mind.cm_skills && user.mind.cm_skills.engineer < SKILL_ENGINEER_METAL)
```
Comment (DominikPanic) [inline code/game/objects/items/explosives/mine.dm]:
```suggestion
	if(user.mind?.cm_skills && user.mind.cm_skills.engineer < SKILL_ENGINEER_METAL)
```

File: code/game/objects/items/explosives/mine.dm [hunk 14]
```diff
@@ -94,35 +105,95 @@
...
 			Bumped(A)
 
-/obj/item/explosive/mine/Bumped(mob/living/carbon/human/H)
-	if(!armed || triggered) return
+/obj/item/explosive/mine/Bumped(mob/M)
+	if(!armed || triggered || iscyborg(M))
+		return
 
-	if((istype(H) && H.get_target_lock(iff_signal)) || iscyborg(H)) return
+	if(istype(M, /mob/living/carbon/human) )
```
Comment (DominikPanic) [inline code/game/objects/items/explosives/mine.dm]: Use ishuman

File: code/game/objects/items/explosives/mine.dm [hunk 15]
```diff
@@ -94,35 +105,95 @@
...
+	if(trigger_type != "Monitoring")
+		triggered = 1
 	playsound(loc, 'sound/weapons/mine_tripped.ogg', 25, 1)
-	trigger_explosion()
+	trigger_explosion(M)
 
 //Note : May not be actual explosion depending on linked method
-/obj/item/explosive/mine/proc/trigger_explosion()
+/obj/item/explosive/mine/proc/trigger_explosion(mob/M)
 	set waitfor = 0
```
Comment (DominikPanic) [inline code/game/objects/items/explosives/mine.dm]: This is innecessary as you aren't sleeping here

File: code/game/objects/items/explosives/mine.dm [hunk 16]
```diff
@@ -170,4 +243,43 @@
 		return
 
 	if(linked_claymore && ismob(A))
-		linked_claymore.Bumped(A)
\ No newline at end of file
+		linked_claymore.Bumped(A)
+
+/obj/item/explosive/mine/proc/mine_alert(mob/M)
+	set waitfor = FALSE
```
Comment (DominikPanic) [inline code/game/objects/items/explosives/mine.dm]: Why the waitfor? I don't see  any sleeps

File: code/modules/projectiles/updated_projectiles/guns/flamer.dm [hunk 17]
```diff
@@ -486,17 +487,20 @@
...
-						else
-							C.adjust_fire_stacks(fire_stacks)
-							var/armor_block = C.run_armor_check("chest", "energy")
-							C.apply_damage(fire_damage, BURN, null, armor_block)
-							C.IgniteMob()
-							C.visible_message("<span class='danger'>[C] bursts into flames!</span>","[isxeno(C)?"<span class='xenodanger'>":"<span class='highdanger'>"]You burst into flames!</span>")
+
+/obj/flamer_fire/proc/burn_carbon(loc, fire_stacks, fire_damage)
+	var/mob/living/C
+	if(fire_stacks || fire_damage)
```
Comment (DominikPanic) [inline code/modules/projectiles/updated_projectiles/guns/flamer.dm]: Early return

**Comment by DominikPanic** [comment]
This PR hasn't been touched for eternity. Feel free to make a new one with all of the aforementioned changes.

**Pull request closed without merging**